## Table of Contents

1. Drop Columnds
2. Rename Columns
3. Fix Index
4. Make a Subset
5. Change Data type
6. Create and use a data dictionary
7. Export

### In Exercise Commands

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [3]:
 # Set path 
path = r"C:\Users\Jenna\Desktop\Data Course\Part4_Instacart_analysis\02_Data\Original_Data"

In [4]:
# import data
df_prods = pd.read_csv(os.path.join(path, "products.csv"),index_col = False)
df_ords = pd.read_csv(os.path.join(path, "orders.csv"),index_col = False)

In [5]:
# drop column
df_ords.drop(columns= ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [6]:
# Set new data fram with dropped column
df_ords = df_ords.drop(columns = ['eval_set'])

In [7]:
# rename dow column
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [8]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [9]:
# re-categorize order_id as string
df_ords['order_id'] =df_ords['order_id'].astype('str')

In [10]:
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [11]:
# import departments csv
df_dep = pd.read_csv(os.path.join(path, 'departments.csv'), index_col =  False)

In [12]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [13]:
# Transpose department df
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [14]:
df_dep_t = df_dep.T

In [15]:
# add index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
# declare new header as row 1
new_header = df_dep_t.iloc[0]

In [17]:
# declare a new df that is just row 1 onwards
df_dep_t_new = df_dep_t[1:]

In [18]:
# label the header as our declared header from above 
df_dep_t_new.columns = new_header

In [19]:
# make this into a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [20]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [21]:
# pull info from data dictionary
print(data_dict.get('19'))

{'department': 'snacks'}


In [22]:
# create a subset dataframe that is just data from the snacks department
df_snacks = df_prods[df_prods['department_id']==19]

In [23]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# TASK START Exercise 4.4

### Question 1: Change another variable in orders df to suitable form

In [24]:
# user_id should also be treated as a string as it is not quantitative

df_ords['user_id'] = df_ords['user_id'].astype('str')

### Question 2: Rename a variable in orders dataframe

In [25]:
# the order number seems to correspond to the user_id, I guess that is what I would change
df_ords.rename(columns = {'order_number' : 'user_order_number' }, inplace =True )

### Question 3: Find the frequency of the hours of the day for orders

In [26]:
# Bringing the count of orders for each hour of the day
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

### Question 3: The most orders occured at 10,11,15 and the least between 2 and 4

### Question 4: Print value 4 for department_id from data dictionary

In [27]:
print (data_dict.get('4'))

{'department': 'produce'}


### Question 4: department_id 4 is produce

### Question 5: Create subset for breakfast items

In [28]:
#Step 1: looking at the data dictionary I have printed in the exercise, brekfast department_id is 14

#Step 2: declare a new dataframe that is equal to only the data from products where department_id =14

df_breakfast = df_prods[df_prods['department_id'] == 14]

#Check that it is correct

df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

### Question 6: Create a subset for dinner Items

In [29]:
# Step 1. using the data dictionary, the key/values we need are
# alcohol : 5
# deli : 20
# beverages : 7
# meat/seafood: 12

# step 2 declare a list with the products described (optional: useful if list may be needed again)

dinner = [5, 20, 7, 12]

# step 3 declare a new dataframe containing all the products from our dinner list
 
df_dinner = df_prods.loc[df_prods['department_id'].isin(dinner)]

# step 4 check new data frame

df_dinner.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

### Question 7: Count the rows of the dinner subset

In [30]:
df_dinner.shape

(7650, 5)

### Question 7: there are 7650 rows in the dinner subset dataframe

### Question 8: extract all the information about user 1

In [31]:
# declare a dataframe that contains a subset of information only about user 1, user_id was made a string and needs quotes
df_user1 = df_ords[df_ords['user_id']== '1']

# check

df_user1.head()

order_id user_id  user_order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1                  1                   2                  8   
1  2398795       1                  2                   3                  7   
2   473747       1                  3                   3                 12   
3  2254736       1                  4                   4                  7   
4   431534       1                  5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Question 9: Provide basic stats about user 1

In [32]:
# the describe function will give us stats about user 1 using the user 1 subset

df_user1.describe()

user_order_number  orders_day_of_week  order_hour_of_day  \
count          11.000000           11.000000          11.000000   
mean            6.000000            2.636364          10.090909   
std             3.316625            1.286291           3.477198   
min             1.000000            1.000000           7.000000   
25%             3.500000            1.500000           7.500000   
50%             6.000000            3.000000           8.000000   
75%             8.500000            4.000000          13.000000   
max            11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Question 10/11 Export the new ords dataframe and the transcribe department dataframe

In [35]:
# declare a new path for ease
pathexport = r"C:\Users\Jenna\Desktop\Data Course\Part4_Instacart_analysis\02_Data\Prepared_Data"

# export ords

df_ords.to_csv(os.path.join(pathexport, 'orders_wrangled.csv'))

# export transcribed deparments

df_dep_t_new.to_csv(os.path.join(pathexport, 'deparments_wrangled.csv'))